cmd/sidecar/main.go

```diff
@@ -0,0 +1,233 @@
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	apiv1alpha1 "github.com/percona/percona-server-mysql-operator/api/v1alpha1"
+	"github.com/percona/percona-server-mysql-operator/pkg/mysql"
+	"github.com/pkg/errors"
+)
+
+var log = logf.Log.WithName("sidecar")
+
+type BackupConf struct {
+	Destination string                        `json:"destination"`
+	Storage     apiv1alpha1.BackupStorageSpec `json:"storage"`
+}
+
+func main() {
+	opts := zap.Options{Development: true}
+	logf.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
+
+	mux := http.NewServeMux()
+
+	mux.HandleFunc("/health/", func(w http.ResponseWriter, req *http.Request) {
+		fmt.Fprintf(w, "OK")
+	})
+	mux.HandleFunc("/backup/", backupHandler)
+	mux.HandleFunc("/logs/", logHandler)
+
+	log.Info("starting http server")
+	log.Error(http.ListenAndServe(":6033", mux), "http server failed")
+}
+
+func getNamespace() (string, error) {
+	ns, err := os.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+	if err != nil {
+		return "", errors.Wrap(err, "read namespace file")
+	}
+
+	return string(ns), nil
+}
+
+func xtrabackupArgs() []string {
+	return []string{
+		"--backup",
+		"--stream=xbstream",
+		fmt.Sprintf("--user=%s", os.Getenv("BACKUP_USER")),
+		fmt.Sprintf("--password=%s", os.Getenv("BACKUP_PASSWORD")),
+	}
+}
+
+func xbcloudArgs(conf BackupConf) []string {
+	args := []string{"put", "--md5", "--parallel=10"}
+
+	switch conf.Storage.Type {
+	case apiv1alpha1.BackupStorageGCS:
+		args = append(
+			args,
+			[]string{
+				"--storage=google",
+				fmt.Sprintf("--google-bucket=%s", conf.Storage.GCS.Bucket),
+				fmt.Sprintf("--google-endpoint=%s", conf.Storage.GCS.EndpointURL),
+				fmt.Sprintf("--google-access-key=%s", conf.Storage.GCS.AccessKey),
+				fmt.Sprintf("--google-secret-key=%s", conf.Storage.GCS.SecretKey),
+			}...,
+		)
+	case apiv1alpha1.BackupStorageS3:
+		args = append(
+			args,
+			[]string{
+				"--storage=s3",
+				fmt.Sprintf("--s3-bucket=%s", conf.Storage.S3.Bucket),
+				fmt.Sprintf("--s3-region=%s", conf.Storage.S3.Region),
+				fmt.Sprintf("--s3-endpoint=%s", conf.Storage.S3.EndpointURL),
+				fmt.Sprintf("--s3-access-key=%s", conf.Storage.S3.AccessKey),
+				fmt.Sprintf("--s3-secret-key=%s", conf.Storage.S3.SecretKey),
+			}...,
+		)
+	case apiv1alpha1.BackupStorageAzure:
+		args = append(
+			args,
+			[]string{
+				"--storage=azure",
+				fmt.Sprintf("--azure-storage-account=%s", conf.Storage.Azure.StorageAccount),
+				fmt.Sprintf("--azure-container-name=%s", conf.Storage.Azure.ContainerName),
+				fmt.Sprintf("--azure-endpoint=%s", conf.Storage.Azure.EndpointURL),
+				fmt.Sprintf("--azure-access-key=%s", conf.Storage.Azure.AccessKey),
+			}...,
+		)
+	}
+
+	args = append(args, conf.Destination)
+
+	return args
+}
+
+func backupHandler(w http.ResponseWriter, req *http.Request) {
```

Comment:
in the log we need to print the command which is used for the backup

Reply:
https://github.com/percona/percona-server-mysql-operator/pull/77/commits/4bc58dc149589b1d7608fb7a9cf778bfe0257644